### PR TITLE
Array method object syntax

### PR DESCRIPTION
**Pull request by chhatch**
Closes - [Array methods using object-syntax](https://github.com/elite-libs/rules-machine/issues/45)

- [x] map
- [x] filter
- [x] every
- [x] some
- [x] find

### CODE REVIEW

**Comment by justsml** [comment]
Dude! F*cking wow 🤘

File: src/expression-language/index.ts [hunk 1]
```diff
@@ -423,16 +423,20 @@ export const ruleExpressionLanguage = function(
     FILTER_VALUES: filterValues,
     INCLUDES_VALUES: filterValues,
     GET: (arg1, arg2) => {
-      const key = string(arg1());
+      // this needs to accept a string or number
+      const key = arg1();
```
Comment (justsml) [inline src/expression-language/index.ts]: Can we make sure it either gets run through `number()` or `string()`? Consistency, etc.

File: src/expression-language/index.ts [hunk 2]
```diff
@@ -423,16 +423,20 @@ export const ruleExpressionLanguage = function(
...
       const inputObj = obj(arg2());
 
+      // @ts-expect-error
       return get(inputObj, key);
     },
     PUT: (arg1, arg2, arg3) => {
-      const key = string(arg1());
+      // this needs to accept a string or number
+      const key = arg1();
```
Comment (justsml) [inline src/expression-language/index.ts]: See above. 🤘

File: src/index.ts [hunk 3]
```diff
@@ -75,76 +76,66 @@ export function ruleFactory<
...
-              rule: term,
-              stepRow,
-              stepCount,
-            });
-          }
-        }
-      } else {
-        // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
-        throw new Error(`Invalid term: ${term}`);
+      if (typeof term !== 'string') throw new Error(`Invalid term: ${term}`);
```
Comment (justsml) [inline src/index.ts:90]: To fix this, we could make the term optional - though maybe not what we want.
```ts
    const parser = init(ruleExpressionLanguage, (term?: string) => {
```
Or, add a bit more defensive posture, by checking the length:
```ts
      if (typeof term !== 'string' || term.length < 1) throw new Error(`Invalid term: ${term}`);
```

File: src/index.ts [hunk 4]
```diff
@@ -312,18 +325,19 @@ export function ruleFactory<
...
-      stepRow: number;
-      input: TInput;
-      rule: string | string[] | Rule;
-      ignoreMissingKeys?: boolean;
+      stepRow: number
+      input: TInput
+      rule: string | string[] | Rule
+      ignoreMissingKeys?: boolean
     }): RuleResult {
+      // checking only the first rule seems unsafe
```
Comment (justsml) [inline src/index.ts:388]: It's about time this index zero check gets more robust. 😅 

What do you think about a composable utility, roughly:

```ts
const validateRules = {
  isString: rule => typeof rule === 'string' && rule.length > 4,
  isStringOrStringList: rule => Array.isArray(rule) ? rule.every(validateRules.isString) : validateRules.isString(rule),
  isLogicalRule: rule => rule != null && ('and' in rule || 'or' in rule),
  isStructuralRule: rule => rule != null && ('if' in rule || 'and' in rule || 'or' in rule || 'map' in rule || 'reduce' in rule || 'filter' in rule || 'return' in rule),
  // TODO: add more combined rules to express higher order structures, constraints, etc
}
```

File: src/index.ts [hunk 5]
```diff
@@ -406,17 +416,26 @@ export function extractValueOrLiteral<
...
-      else?: Rule;
-    }
+    if: And | Or | string
+    then: Rule
+    else?: Rule
+  }
   | {
-      and: Rule[];
-    }
+    and: And
```
Comment (justsml) [inline src/index.ts]: Should the `map` and `try` objects be dedicated interfaces too?

File: src/pr.test.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+describe('array method object syntax', () => {
+  describe('map', () => {
+    it('execute', () => {
+      const rules = [
+        'idList = [12, 34, 56]',
+        {
+          map: 'idList',
+          run: [
+            {
+              if: 'GET(ITER.item, userScore) < 0',
```
Comment (justsml) [inline src/pr.test.ts]: I don't know if I'm 💯% on the `ITER` symbol, 🤔

**Review by justsml — COMMENTED** [review]
Absolutely amazing work on each of these PRs @chhatch!

This is top shelf work 🥃 cheers! 🥇

NIT: Let's see if we can get feedback on the `ITER` symbol this week, I want to build confidence before releasing.

File: src/pr.test.ts (hunk 6 above)
Comment (chhatch) [reply src/pr.test.ts]: I'm sure there's a better name. Using the input context seemed like the easiest way to give the parser access to the array element being mapped.

File: src/expression-language/index.ts (hunk 1 above)
Comment (chhatch) [reply src/expression-language/index.ts]: So, determine the type then run it through one of those?

**Pull request merged**